src/analysis/inner_product_space/adjoint.lean

```diff
@@ -186,6 +186,43 @@ lemma is_adjoint_pair (A : E' →L[ℝ] F') :
 
 end real
 
+section unitary
+
+variables {U : (E →L[𝕜] E)}
+
+/-- Builds a linear equivalence from `U ∈ unitary (E →L[𝕜] E)`. -/
+def linear_equiv_of_unitary (hU : U ∈ unitary (E →L[𝕜] E)) : (E ≃ₗ[𝕜] E) :=
+{ to_fun    := U,
+  map_add'  := map_add U,
+  map_smul' := map_smul U,
+  inv_fun   := (star U : (E →L[𝕜] E)),
+  left_inv  := by { rw function.left_inverse_iff_comp, ext, simp only [id.def, function.comp_app,
+    ← continuous_linear_map.mul_apply, unitary.star_mul_self_of_mem hU,
+    continuous_linear_map.one_apply] },
+  right_inv := by { rw function.right_inverse_iff_comp, ext, simp only [id.def, function.comp_app,
+    ← continuous_linear_map.mul_apply, unitary.mul_star_self_of_mem hU,
+    continuous_linear_map.one_apply] } }
+
+/-- Builds a linear isometric equivalence from `U ∈ unitary (E →L[𝕜] E)`. -/
+def linear_isometry_equiv_of_unitary (hU : U ∈ unitary (E →L[𝕜] E)) : (E ≃ₗᵢ[𝕜] E) :=
+{ to_linear_equiv := linear_equiv_of_unitary hU,
+  norm_map' := λ x : E, by { simp only [linear_equiv_of_unitary, linear_equiv.coe_mk],
+    rw [← sq_eq_sq (norm_nonneg (U x)) (norm_nonneg x), norm_sq_eq_inner,
+      ← continuous_linear_map.adjoint_inner_left, ← continuous_linear_map.mul_apply],
+    rw unitary.mem_iff at hU,
+    rw [← continuous_linear_map.star_eq_adjoint, hU.1, continuous_linear_map.one_apply,
+      inner_self_eq_norm_sq] } }
```

Comment:
I would just build the `linear_isometry_equiv` all at once. There is a coercion to `linear_equiv` anyway. Also, you are in the `continuous_linear_map` namespace, so you don't need to write that long prefix over and over. If you weren't already in that namespace, then just `open continuous_linear_map`. In addition, going through `function.left_inverse_iff_comp` seems like you are taking the long way round. You only want that if you have a "point-free" argument. In this case, it was negated by going back through `id.def, function.comp_app`.
```suggestion
/-- Builds a linear isometric equivalence from `U ∈ unitary (E →L[𝕜] E)`. -/
def linear_isometry_equiv_of_unitary (hU : U ∈ unitary (E →L[𝕜] E)) : (E ≃ₗᵢ[𝕜] E) :=
{ to_linear_equiv :=
  { to_fun    := U,
    map_add'  := map_add U,
    map_smul' := map_smul U,
    inv_fun   := (star U : (E →L[𝕜] E)),
    left_inv  := λ x, by rw [←mul_apply, unitary.star_mul_self_of_mem hU, one_apply],
    right_inv := λ x, by rw [←mul_apply, unitary.mul_star_self_of_mem hU, one_apply] },
  norm_map' := λ x, by {
    rw unitary.mem_iff at hU,
    rw [linear_equiv.coe_mk, ←sq_eq_sq (norm_nonneg (U x)) (norm_nonneg x), norm_sq_eq_inner,
      ←adjoint_inner_left, ←mul_apply, ←star_eq_adjoint, hU.1, one_apply, inner_self_eq_norm_sq] } }
```

Reply:
I appreciate the tips here.  These all seem very clear with hindsight!